File: agent/src/main/java/io/scalac/mesmer/agent/akka/actor/BoundedQueueProxy.java
package io.scalac.mesmer.agent.akka.actor;

import java.util.Collection;
import java.util.Iterator;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

public final class BoundedQueueProxy<E> implements BlockingQueue<E> {

  private final BlockingQueue<E> underlying;

  public boolean getResult() {
    return lastEnqueueResult;
  }

  public void setResult(boolean lastEnqueueResult) {
    this.lastEnqueueResult = lastEnqueueResult;
  }

  private boolean lastEnqueueResult;

  public BoundedQueueProxy(BlockingQueue<E> underlying) {
    this.underlying = underlying;
  }

  @Override
  public boolean add(E e) {
    lastEnqueueResult = underlying.add(e);
    return lastEnqueueResult;
  }

  @Override
  public boolean offer(E e) {
    lastEnqueueResult = underlying.offer(e);
    return lastEnqueueResult;
  }

  @Override
  public void put(E e) throws InterruptedException {
    underlying.put(e);
  }

  @Override
  public boolean offer(E e, long timeout, TimeUnit unit) throws InterruptedException {
    lastEnqueueResult = underlying.offer(e, timeout, unit);
    return lastEnqueueResult;
  }

  @Override
  public E take() throws InterruptedException {
    return underlying.take();
  }

  @Override
  public E poll(long timeout, TimeUnit unit) throws InterruptedException {
    return underlying.poll(timeout, unit);
  }

  @Override
  public int remainingCapacity() {
    return underlying.remainingCapacity();
  }

  @Override
  public boolean remove(Object o) {
    return underlying.remove(o);
  }

  @Override
  public boolean contains(Object o) {
    return underlying.contains(o);
  }

  @Override
  public int drainTo(Collection<? super E> c) {
    return underlying.drainTo(c);
  }

  @Override
  public int drainTo(Collection<? super E> c, int maxElements) {
    return underlying.drainTo(c, maxElements);
  }

  @Override
  public E remove() {
    return underlying.remove();
  }

  @Override
  public E poll() {
    return underlying.poll();
  }

  @Override
  public E element() {
    return underlying.element();
  }

  @Override
  public E peek() {
    return underlying.element();
  }

  @Override
  public int size() {
    return underlying.size();
  }

  @Override
  public boolean isEmpty() {
    return underlying.isEmpty();
  }

  @Override
  public Iterator<E> iterator() {
    return underlying.iterator();
  }

  @Override
  public Object[] toArray() {
    return underlying.toArray();
  }

  @Override
  public <T> T[] toArray(T[] a) {
    return underlying.toArray(a);
  }

  @Override
  public boolean containsAll(Collection<?> c) {
    return underlying.containsAll(c);
  }

  @Override
  public boolean addAll(Collection<? extends E> c) {
    return underlying.addAll(c);
  }

  @Override
  public boolean removeAll(Collection<?> c) {
    return underlying.removeAll(c);
  }

  @Override
  public boolean retainAll(Collection<?> c) {
    return underlying.retainAll(c);
  }

  @Override
  public void clear() {
    underlying.clear();
  }

  @Override
  public boolean equals(Object o) {
    if (o instanceof BoundedQueueProxy) {
      BoundedQueueProxy<E> bqp = (BoundedQueueProxy<E>) o;
      return bqp.lastEnqueueResult == this.lastEnqueueResult && underlying.equals(o);
    }
    return false;
  }

  @Override
  public int hashCode() {
    return underlying.hashCode();
  }
}

File: agent/src/main/scala/io/scalac/mesmer/agent/util/BoundedQueueProxy.scala
//package io.scalac.mesmer.agent.util
//
//import java.util
//import java.util.concurrent.{ BlockingQueue, TimeUnit }
//
//final class BoundedQueueProxy[E](underlying: BlockingQueue[E]) extends BlockingQueue[E] {
//  private var lastEnqueueResult: Boolean = true
//
//  def add(e: E): Boolean = {
//    lastEnqueueResult = underlying.add(e)
//    lastEnqueueResult
//  }
//
//  def offer(e: E): Boolean = {
//    lastEnqueueResult = underlying.offer(e)
//    lastEnqueueResult
//  }
//
//  def put(e: E): Unit = underlying.put(e)
//
//  def offer(e: E, timeout: Long, unit: TimeUnit): Boolean = {
//    lastEnqueueResult = underlying.offer(e, timeout, unit)
//    lastEnqueueResult
//  }
//
//  def take(): E = underlying.take()
//
//  def poll(timeout: Long, unit: TimeUnit): E = underlying.poll(timeout, unit)
//
//  def remainingCapacity(): Int = underlying.remainingCapacity()
//
//  def remove(o: Any): Boolean = underlying.remove(o)
//
//  def contains(o: Any): Boolean = underlying.contains(o)
//
//  def drainTo(c: util.Collection[_ >: E]): Int = underlying.drainTo(c)
//
//  def drainTo(c: util.Collection[_ >: E], maxElements: Int): Int = underlying.drainTo(c, maxElements)
//
//  def remove(): E = underlying.remove()
//
//  def poll(): E = underlying.poll()
//
//  def element(): E = underlying.element()
//
//  def peek(): E = underlying.peek()
//
//  def size(): Int = underlying.size()
//
//  def isEmpty: Boolean = underlying.isEmpty
//
//  def iterator(): util.Iterator[E] = underlying.iterator()
//
//  def toArray: Array[AnyRef] = underlying.toArray()
//
//  def toArray[T](a: Array[T]): Array[T] = null
//
//  def containsAll(c: util.Collection[_]): Boolean = underlying.containsAll(c)
//
//  def addAll(c: util.Collection[_ <: E]): Boolean = underlying.addAll(c)
//
//  def removeAll(c: util.Collection[_]): Boolean = underlying.removeAll(c)
//
//  def retainAll(c: util.Collection[_]): Boolean = underlying.retainAll(c)
//
//  def clear(): Unit = underlying.clear()
//}
